Add windows font fallbacks

davidhewitt opened the pull request:
This is the first step for #3082

At the moment Alacritty on windows does not use any font fallbacks, so glyphs for emoji cannot be found. I suspect there are many other cases where lack of font fallbacks causes glyphs to be missing.

By adding font fallbacks, we now can rasterize (downsampled) greyscale emoji. They don't look amazing, but hey, something is better than nothing!

The remaining step to add full emoji support is to implement color font rasterizing, which will be a much bigger change and might need upstream support in the `dwrote` crate.

Perhaps others may have ideas how to improve the render quality of these greyscale emojis too.

If this PR looks like the right direction to be heading, there's a couple of things for me to finish off:
- [x] Remove hard-coding of "en-us" locale
- [x] Try to find fallback font in loaded fonts rather than always creating fresh

Possible bugs from testing:
- [x] ~~Figure out how to size emoji correctly - downsample?~~
- [x] ~~Investigate emoji paste issues?~~ (Upstream bug https://github.com/microsoft/terminal/issues/1503)
- [x] ~~Investigate occasional missing glyph~~ (Sounds a lot like upstream bug https://github.com/microsoft/terminal/issues/386)

Before:

![image](https://user-images.githubusercontent.com/1939362/71518263-e7b1e180-28a9-11ea-8be5-cf41af80e320.png)

After:

![image](https://user-images.githubusercontent.com/1939362/71518208-ac171780-28a9-11ea-891e-2c5c6202d905.png)

Fixes #3215.

kchibisov commented:
@davidhewitt could you show how glyphs looks like without downsampling on windows?

davidhewitt commented:
If I don't downsample, then the glyphs are sharp. But their width is much larger than for my current font (Cascadia Code), so we end up with alignment issues:

![image](https://user-images.githubusercontent.com/1939362/71542246-ec36d280-295b-11ea-8957-4b746d519157.png)

I've highlighted one smiley to help show the size issue. Alacritty is drawing a highlight rect around what it thinks is two glyphs in width (because the emoji are double-width glyphs), but it doesn't quite capture the smiley face properly. Notice the right edge is cut off. (I think Alacritty _is_ actually rendering the rest of the glyph, but it's black on black background so we can't see it.)

davidhewitt commented:
In Windows Terminal the alignment looks correct, though I don't see any downsampling code. However, its rendering is slightly different to Alacritty's. I think it renders directly to the window surface rather than rasterizing glyphs to bitmaps first as Alacritty does.

![image](https://user-images.githubusercontent.com/1939362/71542488-f27a7e00-295e-11ea-93f3-1af84fcf6eeb.png)

kchibisov commented:
>Alacritty is drawing a highlight rect around what it thinks is two glyphs in width (because the emoji are double-width glyphs), but it doesn't quite capture the smiley face properly.

Emojis are not two width glyphs in the context of terminal app, their just glyphs with width defined unicode-width-crate(wcwidth in other terminals, etc), so they can be with width equaled to 1.

>I've highlighted one smiley to help show the size issue.

Ehm, it's a bit strange to be honest, so rather then using downsampling you should understand why they're so much to the right. On Linux/BSD downsampling exists to solve a different issue, since emojis on the these platforms are not scaled at all (you're getting for example 138x128 pixels bitmap for every font size and should scale it yourself).

I don't like the windows terminal approach tbh, however I feel like it's just their rendering.

kchibisov commented:
>The remaining step to add full emoji support is to implement color font rasterizing, which will be a much bigger change and might need upstream support in the dwrote crate.

Oh, could you go into more details why we can't just implement color font support with what we have right now? I'm not forcing you to add it here, but just wondering.

davidhewitt commented:
> rather then using downsampling you should understand why they're so much to the right.

I query the metrics for each glyph from the font face returned from the fallback API. IIRC, for the highlighted smiley the advance is expected to be about 25 pixels, but Alacritty is only giving the glyph 22 pixels of space (i.e. two monospace columns). All layout is done from the left, which is why the glyph appears to be off to the right.

> Oh, could you go into more details why we can't just implement color font support with what we have right now?

Starting from this line is where Windows Terminal adds color emojis:

https://github.com/microsoft/terminal/blob/c274b38dccc2abdd1ea2fd5a27a3bd2e391e0bcb/src/renderer/dx/CustomTextRenderer.cpp#L304

There's a number of pieces needed for us to reproduce all that:

- We need access to the DirectWrite factory pointer, but that's currently not exposed by `dwrote`.
- We also then need to cast that pointer to the COM interface `IDWriteFactory4`, which isn't even exposed in `winapi`. So we'd need to define that and a few related interfaces ourselves, and then upstream them.
- Once we've got all that, we're rasterizing to bitmaps instead of the window, so we need to pull in WIC to create bitmap render targets for Direct2d. So even the render code is more complicated than what Windows Terminal is doing.

All of that stuff is also going to add a whole load of `unsafe` calls to various C++ APIs, so I thought we should pause half-way with greyscale rendering. (In the interim I wanted to look at what I might upstream to `dwrote` or `winapi`.)

kchibisov commented:
I'll have an access to a Windows machine for a next few days, so I'll try to play with this fallbacks and "slightly big" emojis thing.

kchibisov commented:
You should sort out locale issue, otherwise it's nearly impossible for me to test it. I've changed my locale to much my system locale (hardcode thing), however it still don't work as expected, since I'm getting missing glyphs symbols, followed by emoji in some cases in other just missing glyph symbol.

I also had to install wsl, since emojis wasn't working for me in cmd.exe and powershell (Windows was just removing them from paste!!!!). At least I can say that CJK glyphs are seems to work fine now. (A small win?)

@davidhewitt Do I need something special to make things work like on your screenshot?

davidhewitt commented:
I've added a commit which uses the correct locale.

> Do I need something special to make things work like on your screenshot?

Not as far as I'm aware. Are you using ConPTY + Windows 10? WinPTY + emojis doesn't work for me in testing.

kchibisov commented:
Yeah, ConPty + Windows 10. Windows just decides to strip emojis sometime from input and paste random things in alacritty.

davidhewitt commented:
> Yeah, ConPty + Windows 10. Windows just decides to strip emojis sometime from input and paste random things in alacritty.

Weird, that would be worth investigating as a separate issue perhaps. Might be a clipboard bug our side, or ConPTY bug to report upstream. Paste _seems_ to be working for me. For now I've just manually created a text file in VSCode which has a bunch of emoji in it, and just displaying that with `cat`.

> since I'm getting missing glyphs symbols, followed by emoji in some cases in other just missing glyph symbol.

I _occasionally_ get a missing glyph symbol which disappears on a resize. To me that suggests an upstream bug in ConPTY. (I'll try to debug that.)

kchibisov commented:
>Weird, that would be worth investigating as a separate issue perhaps. Might be a clipboard bug our side, or ConPTY bug to report upstream. Paste seems to be working for me. For now I've just manually created a text file in VSCode which has a bunch of emoji in it, and just displaying that with cat.

I'll try to create something similar to repro.

>I occasionally get a missing glyph symbol which disappears on a resize. To me that suggests an upstream bug in ConPTY. (I'll try to debug that.)

Yeah, it seems like it. Knowing how alacritty clipboard works it seems like a ConPty issue. The key thing here is that if you copy a missing glyph symbol and paste to e.g. Firefox, it'll still be a missing glyph symbol, but not the underlying text (there should be some char causing missing glyph thing).

davidhewitt commented:
>  I don't see you passing through the font size to dwrote for font fallback, maybe there's something missing?

This is a really interesting point - perhaps I should not be calling `create_font_face` and should instead be doing something slightly different to get a font face of the right size. I'll try to find some time to investigate some point this week...

davidhewitt commented:
Some further investigation and I think I've cracked it.

I'm doing the right thing with the font size: the font size is not relevant to the DirectWrite font face. The font size is only relevant when actually rasterizing the glyph - as part of the `DWRITE_GLYPH_RUN` structure.

I've also found how Windows Terminal solves this issue. Instead of downsampling, it dynamically rescales the font size so that the glyph fits within the cell width. The algorithm is here:

https://github.com/microsoft/terminal/blob/6f7ad99d514c55acc02a8052e208b3a923671c9b/src/renderer/dx/CustomTextLayout.cpp#L425

I'm going to adapt this PR to do the same.

kchibisov commented:
FWIW I don't like the idea of using unicode width crate to make things fit into cell. Just let things overlap, it's not a problem tbh, this 1 width emojis are just bad and you'll be consistent with out other backends. So if you want to follow windows terminal here, just don't use unicode-width to solve problem...

chrisduerr commented:
> FWIW I don't like the idea of using unicode width crate to make things fit into cell. Just let things overlap, it's not a problem tbh, this 1 width emojis are just bad and you'll be consistent with out other backends. So if you want to follow windows terminal here, just don't use unicode-width to solve problem...

Yeah, this is true. Since I've just recently updated our unicode-width crate to the latest version of the standard, I can confidently say that non-combined glyphs all have the correct width in Alacritty. So any width mismatch between OS and Alacritty's width function should be an indication of a bug in the OS' library.

davidhewitt commented:
I've tidied up the remaining pieces on this branch. To remove the `unicode-width` dependency, I'm now going for a much simpler downsizing rule:

- If the glyph from the fallback font is wider than two grid cells, shrink the font size for that glyph so the glyph fits neatly in two grid cells.
- Don't scale anything else.

Hopefully you agree this is a reasonable compromise between keeping the rendering tidy and avoiding the `unicode-width` dependency.

Attached is a screenshot of the greyscale rendering now:

![image](https://user-images.githubusercontent.com/1939362/71664408-c2f9b600-2d50-11ea-8a6f-38c0672747d9.png)

The remaining problems @kchibisov was encountering (paste & missing glyphs) appear to be upstream ConPTY issues; I may look into them but that'll be upstream bugs.

In summary, this PR is feature-complete as far as I'm concerned. I plan to follow up with a separate one in the future with color rendering for emojis once I figure out the best places to put all the pieces.

kchibisov commented:
>I've tidied up the remaining pieces on this branch. To remove the unicode-width dependency, I'm now going for a much simpler downsizing rule:
>
>    If the glyph from the fallback font is wider than two grid cells, shrink the font size for that glyph so the glyph fits neatly in two grid cells.
    Don't scale anything else.
>
>Hopefully you agree this is a reasonable compromise between keeping the rendering tidy and avoiding the unicode-width dependency.

Yeah, the picture and the idea LGTM.

>The remaining problems @kchibisov was encountering (paste & missing glyphs) appear to be upstream ConPTY issues; I may look into them but that'll be upstream bugs.

Could you check that our Windows clipboard is not at fault here? Like making it load and the instantly store to it and paste back to some app? If you need help on how to do it let me know.

davidhewitt commented:
> Could you check that our Windows clipboard is not at fault here? Like making it load and the instantly store to it and paste back to some app?

So I did some quick testing and found the following:
- Copying from Alacritty is fine. I can paste the copied emojis into VSCode no problem.
- Powershell's built-in paste works correctly. This is the paste behavior which I saw was working.
- Alacritty's paste action (Control + Shift + V) does not work correctly. This is the broken paste behaviour you're seeing.

So there _is_ possibly an issue with our clipboard code. I wonder if it's a wider issue beyond just emojis (i.e. might be a problem with a number of unicode glyphs).

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,33 +13,207 @@
+impl DirectWriteRasterizer {
+    fn rasterize_glyph_with_fallback(&mut self, glyph: GlyphKey) -> Result<RasterizedGlyph, Error> {
+        self.rasterize_glyph(glyph).or_else(|error| {
+            if let Error::MissingGlyph(_) = &error {
```
It's bad style to have two levels of indentation for the entire method. These first two conditions should likely be combined for an early return.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,33 +13,207 @@
+        self.rasterize_glyph(glyph).or_else(|error| {
+            if let Error::MissingGlyph(_) = &error {
+                // Try to get a fallback font
+                let font = &self.fonts.get(&glyph.font_key).unwrap().font;
```
Is this `unwrap` safe? It doesn't look like something we should be doing.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,33 +13,207 @@
+                        dwrote::NumberSubstitution::new(
+                            dwrite::DWRITE_NUMBER_SUBSTITUTION_METHOD_NONE,
+                            &locale,
+                            false,
```
Why clone locale earlier and then pass a reference here, instead of just putting this outside of the `map_characters` function and then not requiring to clone later anymore?

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,33 +13,207 @@
+                let utf16_codepoints = encode_utf16(glyph.c);
+                let num_codepoints = utf16_codepoints.len() as u32;
+
+                let fallback = dwrote::FontFallback::get_system_fallback().unwrap();
```
Another `unwrap` where I cannot tell if it's safe or not.

Generally we should prefer `except` over `unwrap` too, even if it's mostly safe. Just so we can find the error with stripped info without any issue.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,33 +13,207 @@
+                    font.weight(),
+                    font.style(),
+                    font.stretch(),
+                );
```
There's so much identation here that it becomes difficult to understand what exactly is going on in this massive statement.

Could we maybe extract some statements, like the `from_text_and_muber_subst` call? Possibly also the `NumberSubstitution` and `TextAnalysisSource`.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,33 +13,207 @@
+                    font.stretch(),
+                );
+
+                if let Some(mapped_font) = fallback_font.mapped_font {
```
What if `mapped_font` is `None`? When would this be the case? Currently we explode right now, so we might as well make that explicit by returning the error directly from this point, rather than making it look like we're performing error correction?

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,33 +13,207 @@
+                    let mut fallback_glyph = glyph;
+                    fallback_glyph.font_key = fallback_font_key;
+
+                    return self.rasterize_fallback_glyph(fallback_glyph, glyph);
```
Hopefully explicit return stataments shouldn't be required anymore when the indentation level is fixed. This kinda makes the control flow messy and hard to understand.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,33 +13,207 @@
-impl crate::Rasterize for DirectWriteRasterizer {
-    type Err = Error;
+impl DirectWriteRasterizer {
+    fn rasterize_glyph_with_fallback(&mut self, glyph: GlyphKey) -> Result<RasterizedGlyph, Error> {
```
I'm not sure about the naming here, since we want to always use this method for rasterizing the glyph basically.

So I feel like `rasterize_glyph` would be the correct name and the `rasterize_glyph` method should be renamed instead (since it's basically just a part of the glyph rasterization process).

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,33 +13,207 @@
+        // cell width.
+        //
+        // We'll only do this for any glyph > 2 cells wide, downsizing it to be just 2 cells wide.
+        // This should help keep things tidy without resizing many glyphs needlessly.
```
Isn't a halfwidth font just as likely to exceed the width as a fullwidth one?

It seems a bit strange that the widths would differentiate so strongly, since our fonts sizes are specified in points.

davidhewitt replied:
> Isn't a halfwidth font just as likely to exceed the width as a fullwidth one?

Yes, a halfwidth glyph might overflow one cell just as a fullwidth glyph might overflow two cells.

Now I've removed the `unicode-width` dependency I don't see a way to reliably handle the halfwidth case. (I cannot find any DirectWrite API to ask if a codepoint is half- or full-width.)

davidhewitt replied:
> Is there some documentation on this being required, or recommended? Or did you mainly look at the source code of the windows terminal without a proper name?

There's no documentation I've found on it. I think it's just a sad fact that Windows' emoji font has completely different widths to my monospace font. Despite Windows' having selected this font supposedly as a suitable fallback!

This algorithm is what I ended up with after a few varied attempts:

- My initial implementation had no scaling, and I saw glyphs with slight offset issues like in [this comment](https://github.com/alacritty/alacritty/pull/3127#issuecomment-569404916)
- To fix the offset problem I then attempted downscaling [like in the OP](https://github.com/alacritty/alacritty/pull/3127#issue-357228997). TBH these images don't look great. At this point I was already using glyph advance to calculate the factor to downscale by.
- I then decided to peek at what the Windows Terminal was doing. I felt their strategy of scaling the font size could work here too, which is why I then decided to implement the same.

It's not great, but I think it looks better than either the out-of-bounds fonts or the downsampled ones from the OP.

chrisduerr replied:
I personally find it questionable to implement such a thing on Windows, while opting to overlap on other platforms. We should be consistent across platforms.

And as we found out in this PR, the process of downscaling isn't simple either. Scaling the output glyph can easily lead to poor looking fonts so it's not really an option. But scaling the font size can not guarantee that the result actually works properly, especially with fonts that are not strictly monospace (which is likely the case for emojis).

kchibisov replied:
>Neither macOS nor Linux/BSD require scaling fonts after fallback. It's strange to me that this would be required on Windows, without providing a recommended way of dealing with this.

Speaking of Linux here, it does require your app to scale things(fixup_factor) and it can also provide an odd scale factor, since user can just set one in a config. So technically, we're scaling on Linux to a user defined point, so it's just smarter than whatever Windows is providing.

>I personally find it questionable to implement such a thing on Windows, while opting to overlap on other platforms. We should be consistent across platforms.

We will be consistent with such change on all platforms, since our overlapping is related to `wcwidth`, however this PR doesn't speculate it and just kind of using more smarter `downsampling` than for example Linux/BSD impl.

>And as we found out in this PR, the process of downscaling isn't simple either. Scaling the output glyph can easily lead to poor looking fonts so it's not really an option. But scaling the font size can not guarantee that the result actually works properly, especially with fonts that are not strictly monospace (which is likely the case for emojis).

They are full-width fonts, we can perform such scaling only on colored fonts tbh, since we can query such information IIRC.

Anyway, @davidhewitt, could you show me how this string looks like `❄m❄` on your PR?

davidhewitt replied:
> Anyway, @davidhewitt, could you show me how this string looks like `❄m❄` on your PR?

Sure, it looks like this when pasted in (I typed a few spaces after to get the cursor out the way):

![image](https://user-images.githubusercontent.com/1939362/72479306-068dfe80-37ec-11ea-9d4c-ac08a93c2a08.png)

kchibisov replied:
One more note, some odd fonts on Linux like `Inconsolata` are playing really bad with all the other fonts on Linux, so technically even normal chars from other fonts are starts to overlap each other, not speaking of emojis. So technically @chrisduerr is right here, I guess.

davidhewitt replied:
Ok. It looks like the consensus is to just not do the scaling, so I'll remove it now.

kchibisov replied:
>Sure, it looks like this when pasted in (I typed a few spaces after to get the cursor out the way):

The picture is absolutely correct though, we and the most of other terminals, have the same thing on Linux. So now it's just a matter of using the same. However we can implement the same algo for font change on Linux, it should be pretty trivial.

kchibisov replied:
@davidhewitt wait a bit.

davidhewitt replied:
Ok. I'm going to head to bed for the night; will pick this up at a later date when a decision has been made. My personal preference is to have the scaling in so the emojis line up nicely, but I will remove it if that's the consensus from everyone else.

We have some time to mull this over as we need the `dwrote` PRs merged before this is mergable now anyway.

kchibisov replied:
After some discussion on IRC, and further investigation with our Linux/BSD backend, we decided that we shouldn't scale things for now and hold it back until there will be a lot of complains in the future. The algorithm since a bit of fragile tbh, since some font metrics are not that much of reliable, knowing that we can fallback to a glyph from non monospace font and there's always tons of use cases.

Also we do have a font rfc #957 and there's an idea on providing some set of options to a font, so with it you can just change a font size to "bad" fonts in your fallback to better match a primary font, since it's more reliable way to do things than guessing based off some font metrics.

tl;dr You issue should be covered by #957 and we should revert any scaling to match our other backends for now.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,33 +13,207 @@
+            self.rasterize_glyph(scaled_glyph)
+        } else {
+            self.rasterize_glyph(glyph)
+        }
```
Reusing `rasterize_glyph` is nice, but calling it twice here still seems a bit strange to me.

Can't we just mutate `glyph` in the `if` block and then always call `self.rasterize_glyph(glyph)` afterwards?

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,33 +13,207 @@
+        let max_advance = 2.0 * font_metrics.average_advance;
+        let actual_advance = glyph_metrics.average_advance;
+
+        if actual_advance > max_advance {
```
As mentioned in another comment, I don't think this is a reliable way to determine if we need to scale the font.

davidhewitt replied:
> As mentioned in another comment, I don't think this is a reliable way to determine if we need to scale the font.

I agree that it's not perfect, though haven't managed to come up with better. What would you like to see instead?

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,33 +13,207 @@
+        }
+    }
+
+    fn metrics_from_glyph(&self, glyph: GlyphKey) -> Result<Metrics, Error> {
```
```suggestion
    fn glyph_metrics(&self, glyph: GlyphKey) -> Result<Metrics, Error> {
```

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -215,3 +356,51 @@ impl ::std::fmt::Display for Error {
+            _ => Weight::Normal,
+        }
+    }
+}
```
Do we need both here, or is this just for completeness? Generally I think it's nicer to have one 'target' format that is used everywhere and we convert to that as soon as possible.

If we need to pass one of our `Weight`s into the `dwrote` crate and we get a `FontWeight` out of `dwrote`, which we need to convert back to our weight, this is obviously the way to go though.

Same applies for `Slant`/`FontStyle` too of course.

davidhewitt replied:
I've gone for just `Weight` -> `dwrote::FontWeight` etc.

davidhewitt commented:
It seems like the paste issue is likely upstream with ConPTY, as Windows Terminal has the same problem: https://github.com/microsoft/terminal/issues/1503

davidhewitt commented on font/src/directwrite/locale.rs:
```diff
@@ -0,0 +1,13 @@
+use super::winapi::shared::ntdef::LOCALE_NAME_MAX_LENGTH;
+use super::winapi::um::winnls::GetUserDefaultLocaleName;
+
+pub fn get_current_locale() -> String {
```
Yes, I was anticipating it to be more complicated and then it turned out short. I'll move it back into `mod.rs`.

davidhewitt commented on font/src/directwrite/mod.rs:
```diff
@@ -14,32 +14,223 @@
+    GlyphRunAnalysis,
 };
 
+use self::winapi::um::dwrite;
```
Unfortunately these `extern crate` commands aren't done in `lib.rs`, which scopes them to only this module but also means `self` is needed.

I'm actually just going to add a PR to bump the font crate to 2018 edition.

chrisduerr replied:
> I'm actually just going to add a PR to bump the font crate to 2018 edition.

Yeah, I think that's a good idea. I honestly don't even know how 2015 works exactly anymore myself (though I swear that worked before!).

davidhewitt commented on font/src/directwrite/mod.rs:
```diff
@@ -14,32 +14,223 @@
+            if fallback_result.is_ok() {
+                return fallback_result;
+            }
+        }
```
Thanks, that's much nicer!

davidhewitt commented on font/src/directwrite/mod.rs:
```diff
@@ -14,32 +14,223 @@
+            dwrite::DWRITE_NUMBER_SUBSTITUTION_METHOD_NONE,
+            &locale,
+            false,
+        );
```
I don't think we actually need it for what we're doing here; I'm going to submit a quick PR upstream to `dwrote` which allows us to get rid of this here.

davidhewitt commented on font/src/directwrite/mod.rs:
```diff
@@ -14,32 +14,223 @@
+        Ok(fallback_font.mapped_font)
+    }
+
+    fn rasterize_font_char(
```
Yes; I'm struggling to come up with a better name though. Will keep thinking.

chrisduerr replied:
I can see if I can come up with some ideas once all the other problems are resolved. Since it's just a style issue getting all the other stuff figured out first probably makes more sense.

davidhewitt commented on font/src/directwrite/mod.rs:
```diff
@@ -14,32 +14,223 @@
+pub struct TextAnalysisSource {
+    locale: String,
+    length: u32,
+}
```
Will add comment and remove the superfluous `pub`

davidhewitt commented on font/src/directwrite/mod.rs:
```diff
@@ -14,32 +14,223 @@
+    }
+}
+
+pub struct FontEntry {
```
I'll go for `LoadedFont`.

davidhewitt commented:
I fixed a number of the style issues and rebased on master.

This PR won't build at the moment. I opened two PRs against `dwrote` which we will need released before this is mergeable. The code as written in this PR currently assumes both of them:
- https://github.com/servo/dwrote-rs/pull/37
- https://github.com/servo/dwrote-rs/pull/38

chrisduerr commented:
> I fixed a number of the style issues and rebased on master.
>
> This PR won't build at the moment. I opened two PRs against dwrote which we will need released before this is mergeable.

Thanks for keeping us posted.

davidhewitt commented:
Font scaling removed. Hopefully the `dwrote` PRs don't get stuck for too long. The one removing `NumberSubstitution` would especially be desirable to merge so we don't have pointless code kicking around here. The other one is just a convenience method that we could include as a free function here if needed.

davidhewitt commented:
The `dwrote` PRs have now been merged, but not yet released. I've pushed a commit which switches to use the `git` version of `dwrote` (pinned to a specific revision), and rebases on master.

Barring a little bikeshedding on function names (the unresolved comment above), and whether you want to wait for a full `dwrote` release, this is now ready to go.

kchibisov commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,157 @@
-    type Err = Error;
+impl DirectWriteRasterizer {
+    fn rasterize_glyph(&mut self, glyph: GlyphKey) -> Result<RasterizedGlyph, Error> {
+        let loaded_font = &self.fonts.get(&glyph.font_key).ok_or(Error::FontNotLoaded)?;
```
This function looks like "pointer to impl" thing, what do you thing about moving this logic into `get_glyph` method and merge `rasterize_glyph` and `rasterize_font_char` in one function, so it'll solve a problem with naming and also remove one layer of abstraction.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,157 @@
-    fn metrics(&self, key: FontKey, size: Size) -> Result<Metrics, Error> {
-        let font = self.fonts.get(key.token as usize).ok_or(Error::FontNotLoaded)?;
+    fn get_fallback_font(&self, font: &Font, c: char) -> Result<Font, Error> {
+        let fallback = match self.font_fallback.as_ref() {
```
```suggestion
        let fallback = match &self.font_fallback {
```

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,157 @@
+        let length = utf16_codepoints.len() as u32;
+
+        let locale = get_current_locale();
+        let text_analysis_source = dwrote::TextAnalysisSource::from_text(
```
```suggestion
        let text_analysis_source = TextAnalysisSource::from_text(
```

Should probably `use dwrote::TextAnalysisSource` at the top.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,157 @@
+
+        let glyph_index = *font_face
+            .get_glyph_indices(&[c as u32])
+            .first()
```
```suggestion
            .first()
            // DirectWrite returns 0 if the glyph does not exist in the font
            .filter(|glyph_index| glyph_index == 0)
```

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,157 @@
+            glyphCount: 1,
+            glyphIndices: &(glyph_index),
+            glyphAdvances: &(0.0),
+            glyphOffsets: &(offset),
```
What are these parenthesis for?

davidhewitt replied:
Hangover from the previous author's implementation. I'll remove them

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,157 @@
+            0.0,
+            0.0,
+        )
+        .or_else(|_| Err(Error::MissingGlyph(c)))?;
```
Is this error really correct? We shouldn't be returning a missing glyph error if that's not the actual problem.

Same goes for all the other errors in this function following this one.

davidhewitt replied:
I've gone for adding a `DirectWriteError` variant which stores the `HRESULT` error code returned from these functions. We could potentially give the variant a more specific name like `GlyphRasterizationFailed`?

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,157 @@
+        // We assume that all monospace characters have the same width
+        // Because of this we take '!', the first drawable character, for measurements
+        self.glyph_metrics(font_face, '!', size)
+    }
```
In previous versions this was different, but I don't see multiple users of these metric functions anymore.

It seems like font_metrics, glyph_metrics and metrics can all just be merged into one function?

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -107,73 +232,42 @@ impl crate::Rasterize for DirectWriteRasterizer {
-            .create_alpha_texture(dwrote::DWRITE_TEXTURE_CLEARTYPE_3x1, bounds)
-            .or_else(|_| Err(Error::MissingGlyph(glyph.c)))?;
+    fn load_font(&mut self, desc: &FontDesc, _size: Size) -> Result<FontKey, Error> {
+        self.get_matching_face(desc)
```
It seems like this `get_matching_face` function is also only ever called from here? Then we can probably inline it.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -212,3 +306,57 @@ impl ::std::fmt::Display for Error {
+}
+
+impl From<Weight> for FontWeight {
+    fn from(other: Weight) -> FontWeight {
```
```suggestion
    fn from(weight: Weight) -> FontWeight {
```

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -212,3 +306,57 @@ impl ::std::fmt::Display for Error {
+}
+
+impl From<Slant> for FontStyle {
+    fn from(other: Slant) -> FontStyle {
```
```suggestion
    fn from(slant: Slant) -> FontStyle {
```

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -212,3 +306,57 @@ impl ::std::fmt::Display for Error {
+    length: u32,
+}
+
+impl dwrote::TextAnalysisSourceMethods for TextAnalysisSourceData {
```
```suggestion
impl TextAnalysisSourceMethods for TextAnalysisSourceData {
```

Probably should add `use dwrote::TextAnalysisSourceMethods`.

davidhewitt commented:
Thanks for all the feedback. I've made a number of tweaks to address those points, and also rebased on master.

kchibisov commented on font/Cargo.toml:
```diff
@@ -23,4 +23,5 @@ core-graphics = "0.17"
 
 [target.'cfg(windows)'.dependencies]
-dwrote = { version = "0.9.0" }
+dwrote = { git = "https://github.com/servo/dwrote-rs", rev = "c2486e8" }
```
Just to note, I don't really mind in using git dependency on Windows for the time being.

chrisduerr replied:
I do. We should usually stay away from git dependencies because we have no control over those retrospectively. With cargo we know immutability of versions is guaranteed.

It shouldn't be too hard to cooperate a release though.

kchibisov replied:
I know, I'm just saying that it's fine for me to see them on Windows tbh, since it's likely not that popular platform and support/cooperation could be hard. I don't want to see this mess like we had with winit/glutin on Linux tbh/macOS tbh.

chrisduerr replied:
We've probably got over a thousand Windows users, so it shouldn't be taken lightly.

chrisduerr commented:
Clippy and rustfmt also seem to have some remaining complaints. Once those are done I'll give it another review.

davidhewitt commented:
> 
> 
> Mostly looks good from my side, could you provide us with information on how this PR performs on `vtebench unicode-random-write` and also some information on how long does it take to get a fallback during running it?
> 
> ```rust
> let t1 = std::time::Instant::now();
> let fallback_font = self.get_fallback_font(&loaded_font.font, glyph.c).ok_or(err)?;
> let t2 = std::time::Instant::now();
> println!("Time for fallback {:?}", t2 - t1).
> ```

A random sample of the timings for this:

    Time for fallback 62.9µs
    Time for fallback 24.2µs
    Time for fallback 45.1µs
    Time for fallback 40µs
    Time for fallback 92.9µs
    Time for fallback 80.4µs
    Time for fallback 31.4µs
    Time for fallback 120.5µs
    Time for fallback 52.2µs
    Time for fallback 67µs
    Time for fallback 33.8µs
    Time for fallback 40.5µs
    Time for fallback 125.5µs

Also, for `vtebench` with 10MB input, it takes ~17 seconds to render on this branch, versus about 17 seconds on `master` too. For comparison, Windows terminal churns through the same file over about 5 minutes. I suspect `Alacritty` might be skipping frames as it buffers?

With this branch, the vtebench output has a lot more characters rendered as compared to master (e.g. CJK characters). However, the framerate does start noticeably slower than master (but not terrible), and then improves as the fallbacks for missing glyphs in the primary font are cached.

davidhewitt commented:
Fixed clippy errors.

kchibisov commented:
@davidhewitt you should output vtebench's output to a file and then `cat` it. 
so `vtebench unicode-random-write > out.txt` and then `cat out.txt`. Alacritty doesn't skip frames on purpose, it can start skipping them if your fallback is really slow, but we don't have any logic to do that, since fast terminal shouldn't skip frames, etc.

>Also, for vtebench with 10MB input, it takes ~17 seconds to render on this branch, versus about 17 seconds on master too. For comparison, Windows terminal churns through the same file over about 5 minutes. I suspect Alacritty might be skipping frames as it buffers?

how it could be so long. on Linux 500MB of unicode-random-write takes me 4.4s. I feel like this is just throttled by ConPTY/WinPTY. 17s for 10MB is odd.

I'm not sure whether times are good or not, compared to what we have now on Linux/BSD they are terrible, like 5-6x slower on average, but Windows also doesn't help here. I encourage you to take a look whether times for fallback could be any better. Ideally you should have max time of 30-40usec.

davidhewitt commented:
> @davidhewitt you should output vtebench's output to a file and then cat it.
so vtebench unicode-random-write > out.txt and then cat out.txt

This is exactly what I did.

> I'm not sure whether times are good or not, compared to what we have now on Linux/BSD they are terrible, like 5-6x slower on average, but Windows also doesn't help here. I encourage you to take a look whether times for fallback could be any better. Ideally you should have max time of 30-40usec.

Yes, windows performance on Alacritty isn't amazing compared to Linux. I've previously played with this and come to the conclusion that ConPTY is a bottleneck.

WRT making the max time of fallback faster, I can do a few tweaks to cache things like the locale and the font properties to avoid a few Windows API calls. It shaves off a little bit of time.

It would be possible to save a number allocations in the font fallback paths, but I'll need to do some further PRs to dwrote. Using those as well as the caching described, the average seems to be ~30usec, but the max time still occasionally spikes to 80-100usec.

kchibisov commented:
>Yes, windows performance on Alacritty isn't amazing compared to Linux. I've previously played with this and come to the conclusion that ConPTY is a bottleneck.

Well, this is due to Windows, since our internal structure is purely generic and the same for all backends, and we're using the same OpenGL APIs on each platform. So you're basically fighting ConPTY and WinPTY here.

>but the max time still occasionally spikes to 80-100usec.

Oh, the average of 30usec could be quite good actually for Windows. Spikes for 80-100usecs could be due to initial loading of font or some other stuff, like missing glyph, I guess. Well I don't have much to recommend, however caching some computations unlikely dramatically improve something. I'll recommend to check how much time are you spending on some Windows API calls.

You can also take a look on what we've done recently for Linux/BSD for fallbacks here, but it'll unlikely help you. https://github.com/alacritty/alacritty/blob/6b327b6f8f0f308ff8f46cdf551ce0d0f3eda60b/font/src/ft/mod.rs#L617

kchibisov commented:
Just to make it clear, the initial slowness is expected, since our render caching things and probing fallback for 30usec(you have 2ms in total for fallback and other stuff + rendering) hurts, however it should cache all the thing during one run of `unicode-random-write` or something similar to this, I guess.

Anyway, I'm just trying to say, that fallback performance is very important and we should make sure that it is compatible to our other backends(at least Linux/BSD one).

davidhewitt commented:
👍 - I'll go ahead and put together the `dwrote` PR in an evening soon, and push some performance tweaks here.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,159 @@
+        &self,
+        font_face: &FontFace,
+        c: char,
+        size: Size,
```
Super minor thing, but I'd swap these two parameters just because that would be a bit more logical and intuitive.

So `rasterize_glyph(font_face, size, c);`.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,159 @@
+            glyphCount: 1,
+            glyphIndices: &glyph_index,
+            glyphAdvances: &0.0,
+            glyphOffsets: &offset,
```
```suggestion
            glyphOffsets: &GlyphOffset::default(),
```

Think we can simplify things a bit by inlining this here. According to winapi docs, `Default` should be implemented.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,159 @@
+        font_face
+            .get_glyph_indices(&[c as u32])
+            .first()
+            .cloned()
```
Why do we have to `cloned` here? Would it otherwise return a `&u16`? Might be a bit more conventional to just put a `*` before `font_face`? If you've had troubles with `filter`, a `**glyph_index != 0` double deref is sometimes necessary.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -107,73 +231,29 @@ impl crate::Rasterize for DirectWriteRasterizer {
+            Err(err @ Error::MissingGlyph(_)) => {
+                let fallback_font =
+                    self.get_fallback_font(&loaded_font.font, glyph.c).ok_or(err)?;
+                self.rasterize_glyph(&fallback_font.create_font_face(), glyph.c, glyph.size)
```
This is also super minor, but it seems unfortunate that we *just barely* linebreak here.

I'd be fine with renaming this to `fallback` or just `font`. Though `fallback_font` is definitely the best variable name, so I don't mind that too much either.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -186,29 +266,79 @@ pub enum Error {
-            Error::MissingGlyph(ref c) => write!(f, "Glyph not found for char {:?}", c),
-            Error::MissingFont(ref desc) => write!(
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> ::std::fmt::Result {
```
We can move all the `fmt` stuff to imports here. There are no conflicts with anything as far as I can tell.

kchibisov commented:
@davidhewitt Could you ensure that your PR preserves style during fallback? It was mentioned there #3271 , I think it would be nice to match style of original glyph.

davidhewitt commented:
Just a ping to say I haven't forgotten about this; have been very busy!

chrisduerr commented:
@davidhewitt No worries, just let us know if you need any help.

davidhewitt commented:
> @davidhewitt Could you ensure that your PR preserves style during fallback? It was mentioned there #3271 , I think it would be nice to match style of original glyph.

I just checked this quickly and confirmed both red colour and bold are preserved. So I believe this is working.

chrisduerr commented:
> I just checked this quickly and confirmed both red colour and bold are preserved. So I believe this is working.

Fyi you can't do anything about color, but italics and bold are relevant. So the question is if a glyph that is not part of the main font, like '汉' can be stylized bold and italic.

davidhewitt commented:
Yep, is all good: 

![image](https://user-images.githubusercontent.com/1939362/76146271-3a1d3480-6089-11ea-8321-3fcddda4aa03.png)

chrisduerr commented:
Awesome, thanks for testing.

davidhewitt commented:
Rebased on master, included some simple perf optimizations (mostly around doing a bit less text copying of things like the locale) and addressed all style tweaks.

To reduce text copying I also had to adjust dwrote slightly with another PR there. https://github.com/servo/dwrote-rs/pull/41

This PR won't build until that dwrote PR is merged.

chrisduerr commented:
@davidhewitt Is this ready to go as far as you're concerned once that PR is merged?

If so, please request a review from me once that's done, then I can take another peek at the code.

davidhewitt commented:
Yes. (And if you want, just merge this pointing at my dwrote branch for now.)

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,170 @@
+
+        // We assume that all monospace characters have the same width
+        // Because of this we take '!', the first drawable character, for measurements
+        let c = '!';
```
Could we move this closer to where it's actually used?

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -107,73 +242,32 @@ impl crate::Rasterize for DirectWriteRasterizer {
+        let stretch = font.stretch();
+        let key = FontKey::next();
+        self.keys.insert(desc.clone(), key);
+        self.fonts.insert(key, LoadedFont { font_face, family_name, weight, style, stretch });
```
All these variables are defined only to be used in the `LoadedFont` initialization. So they shouldn't be defined as separate variables but we should just inline these in the struct definition.

However since all we seem to be doing is querying fields on one thing to create another, this seems like a prime candidate for an `Into` implementation?

I'd assume there's a reason why we can't just use dwrote's `Font` ourselves, right? Otherwise that would be the obvious solution.

davidhewitt replied:
I am caching all of these things because otherwise it creates a lot of additional directwrite API calls to retrieve this information every time a fallback font lookup is needed (all this information is input to the `map_characters` API call).

This was done in the name of performance.

In addition I think `font.create_font_face()` is likely to be relatively heavyweight so would prefer avoid doing that on every glyph rasterization, which we'd have to do if we just store `dwrote::Font`.

I really like the `Into` idea though.

chrisduerr replied:
That's fair enough, I'm not aware of how many of these methods are basically just a memory access versus an actual computation. `Into` would definitely be a good solution either way.

davidhewitt replied:
A lot of the `dwrote` methods involving `String` have at least one additional conversion going on to convert back and forth from UTF-16 C++ encoding to Rust `String`. Based on experiments I've done, caching all these saves us a number of microseconds, which counts for a lot when the comment above was asking me to try to make perf ~ 20-30 us per fallback lookup!

Will go down the `Into` road.

chrisduerr replied:
Great job actually looking into this, I'll definitely gladly take those free performance improvements.

tinnick commented:
Hi, a random rust noob passing by. Is this issue still in progress? Is there anyway I can help here? I'm Japanese and not being able to render any Japanese characters unless I set the primary character to Japanese is quite annoying. Alacritty is awesome btw :)

kchibisov commented:
>Is this issue still in progress?

Yes it is, as you can see.

>Is there anyway I can help here?

You can test this PR.

davidhewitt commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,170 @@
 };
 
+pub struct LoadedFont {
+    font_face: FontFace,
```
Nice, will do that.

davidhewitt commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,170 @@
 
+pub struct LoadedFont {
+    font_face: FontFace,
+    family_name: String,
```
I re-used `family_name` verbatim from `dwrote`, but I can change this to family if preferred.

chrisduerr replied:
No, that's a reasonable argument. Let's just keep things consistent with dwrote then.

davidhewitt commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,170 @@
     device_pixel_ratio: f32,
+    font_collection: FontCollection,
+    font_fallback: Option<FontFallback>,
+    locale: String,
```
Yes, I am aware that was a drawback of caching it like this. I did this to remove the need to call this every time we search for a font fallback.

Let me see what happens to fallback timings if I revert just this bit...

davidhewitt replied:
Ok, I've reverted just this bit. Btw it looks like `dwrote` also caches this thing:

https://github.com/servo/dwrote-rs/blob/c2486e8db9ef1a6fb51b126ead34d392a5060d82/src/helpers.rs#L14-L21

But that's not necessarily a reason for us to be naughty too. Caching this one thing barely affects performance.

davidhewitt commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,170 @@
 
     fn new(device_pixel_ratio: f32, _: bool) -> Result<DirectWriteRasterizer, Error> {
-        Ok(DirectWriteRasterizer { fonts: Vec::new(), device_pixel_ratio })
+        let locale = get_current_locale();
```
👍 I tend to lean towards field init shorthand for everything but that's just my habit and can undo that here!

davidhewitt commented:
> Is this issue still in progress?

It's very nearly complete - we're just waiting for https://github.com/servo/dwrote-rs/pull/41 to be merged and released, which allows some performance optimisations I'm doing here.

davidhewitt commented:
The dwrote PR got merged so I have rebased this PR and updated the `dwrote` git commit referenced.

Fingers crossed CI is green. It's then your call if we want to merge now or request `dwrote` releases a new version to use instead of a git dependency.

chrisduerr commented:
> It's then your call if we want to merge now or request dwrote releases a new version to use instead of a git dependency.

I'd like to stick to actual releases. Could you maybe talk to the dwrote people and at least say that we have interest in a release? It doesn't matter when, but I think it's good to share that others are waiting.

davidhewitt commented:
👍 see https://github.com/servo/dwrote-rs/issues/44

davidhewitt commented:
`dwrote 0.11` is now released, so I've updated this PR and rebased on master.

kchibisov commented on CHANGELOG.md:
```diff
@@ -33,6 +33,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
+### Added
+
+- Font fallback on Windows
```
Rebase error, should be in `0.5.0-dev`.

damnskippy commented:
Just out of curiosity, any chance this can or will be included in the upcoming release? 
Thanks.

chrisduerr commented:
> Just out of curiosity, any chance this can or will be included in the upcoming release?

The 0.4.2 release will no receive any further updates other than bugfixes for regressions.

davidhewitt commented:
> Rebase error, should be in 0.5.0-dev.

Good catch, fixed. Also squashed all commits.

chrisduerr commented:
@davidhewitt I'll try to get to this as soon as my search effort is done. I have not forgotten about this, just to let you know.

codecat commented:
Seems like there's still some issues with glyph alignment/scaling, at least on the latest commit from this pull request. (Unsure if I should be reporting this though, might already be known. Just figured I'd give some hopefully helpful feedback!)

![image](https://user-images.githubusercontent.com/136534/79125772-01b2f980-7d9f-11ea-86f8-81d37968acae.png)

chrisduerr commented:
> Seems like there's still some issues with glyph alignment/scaling, at least on the latest commit from this pull request.

I don't think there's any plan to fix this if I remember the discussion in this PR correctly. We don't scale fonts on any other platform so it doesn't make sense to start that on Windows.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,168 @@
+/// We use this struct rather than dwrote::Font for performance reasons: the getter methods on
+/// dwrote::Font need to do things like String conversion from C++ widestrings to Rust's
+/// UTF-8 strings.
+pub struct Font {
```
Why is this public?

davidhewitt replied:
Doesn't need to be; will change.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,168 @@
     BitmapBuffer, FontDesc, FontKey, GlyphKey, Metrics, RasterizedGlyph, Size, Slant, Style, Weight,
 };
 
+/// This struct contains the properties from dwrote::Font needed in the rest of this module.
+///
+/// We use this struct rather than dwrote::Font for performance reasons: the getter methods on
+/// dwrote::Font need to do things like String conversion from C++ widestrings to Rust's
+/// UTF-8 strings.
```
This documentation isn't really great from a user perspective. It's extremely verbose and contains a lot of information that isn't really necessary or useful.

```suggestion
/// Cached DirectWrite font.
```

I think this might even be sufficient already. You know that it tries to prevent overhead with loading because it's a cache and it says what it is. No need to know anything about C++ widestring conversions.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,168 @@
+    fonts: HashMap<FontKey, Font>,
+    keys: HashMap<FontDesc, FontKey>,
     device_pixel_ratio: f32,
+    collection: FontCollection,
```
This is the font fallback list, correct? I think `collection` might not be ideal naming since a collection might be all kinds of things. On Linux this is called `fallback_lists` for example (since every style has a different fallback list).

Or is this a collection of all fonts present on the system?

davidhewitt replied:
This is the collection of fonts the use. Usually the system collection - and that's what's loaded in `::new()`.

(The dwrote `FontCollection` is a wrapper type around the [DirectWrite interface](https://docs.microsoft.com/en-gb/windows/win32/api/dwrite/nn-dwrite-idwritefontcollection?redirectedfrom=MSDN))

This used to be called `font_collection` but we took the prefix `font_` off most names in the file.

`dwrite_collection` ?

chrisduerr replied:
Usually when I think collection, I'm directed more torwards `std` than directwrite.

I think `available_fonts` would be a good choice, it seems to correctly represent what is stored.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,168 @@
+    keys: HashMap<FontDesc, FontKey>,
     device_pixel_ratio: f32,
+    collection: FontCollection,
+    fallback: Option<FontFallback>,
```
The naming for this is a bit awkward too. Maybe there's a name that would better describe it than just 'fallback'?

davidhewitt replied:
Similar to above, the name `fallback` leads directly from the directwrite `FontFallback` concept.

`fallback_sequences` ? (Inspired by the description of [FontFallback at Microsoft](https://docs.microsoft.com/en-gb/windows/win32/api/dwrite_2/nn-dwrite_2-idwritefontfallback))

chrisduerr replied:
Seems like it would be `fallback_sequence`, without the `s`?

The word `sequence` seems almost a bit hard to grasp for me, but I think it's a decent choice that should work just fine.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,168 @@
+
+    fn get_glyph_index(&self, face: &FontFace, c: char) -> Result<u16, Error> {
+        let idx = *face
+            .get_glyph_indices(&[c as u32])
+            .first()
+            // DirectWrite returns 0 if the glyph does not exist in the font
+            .filter(|glyph_index| **glyph_index != 0)
```
Is there any way for this to return `0` as the first element of the indices, but then continue with a usable non-zero value later?

davidhewitt replied:
As per the comment, which comes from the [remarks on GetGlyphIndices](https://docs.microsoft.com/en-gb/windows/win32/api/dwrite/nf-dwrite-idwritefontface-getglyphindices#remarks) `0` is the value returned when the glyph is not in the font, so that's definitely the error condition.

If you're concerned about the length - `get_glyph_indices()` will always return a `Vec` of length equal to the input slice - so this vec will always have only one element in it.

Perhaps you prefer this alternative implementation:

```
        match face.get_glyph_indices(&[c as u32]).as_slice() {
            [0] => Err(Error::MissingGlyph(c)),
            [i] => Ok(*i),

            // get_glyph_indices will always return a Vec with length equal to the input slice
            _ => unreachable!()
        }
```

chrisduerr replied:
I'm fine with the implementation, I was just concerned if it might be necessary to check for later elements in the vec.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -13,30 +13,168 @@
+        let text_analysis_source_data = TextAnalysisSourceData { locale: &locale, length };
+        let text_analysis_source = TextAnalysisSource::from_text(
+            Box::new(text_analysis_source_data),
+            Cow::Borrowed(&utf16_codepoints),
```
```suggestion
            Cow::Borrowed(utf16_codepoints),
```

I don't think this reference is necessary?

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -51,9 +189,12 @@ impl crate::Rasterize for DirectWriteRasterizer {
         // We assume that all monospace characters have the same width
-        // Because of this we take '!', the first drawable character, for measurements
-        let glyph_metrics = font.get_design_glyph_metrics(&[33], false);
-        let hmetrics = glyph_metrics.first().ok_or(Error::MissingGlyph('!'))?;
+        // Because of this we take '!', the first drawable character, for horizontal metrics
```
```
        // Since all monospace characters have the same width, we use `!` for horizontal metrics
```

I think this works a bit better, even though it has only slightly been changed in this PR. That way we don't have two sentences that aren't properly delimited with punctuation but still separate sentences.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -69,27 +210,22 @@ impl crate::Rasterize for DirectWriteRasterizer {
 
     fn load_font(&mut self, desc: &FontDesc, _size: Size) -> Result<FontKey, Error> {
-        let system_fc = FontCollection::system();
+        // Fast path if face already loaded
```
```suggestion
        // Fast path if face is already loaded
```

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -107,73 +243,27 @@ impl crate::Rasterize for DirectWriteRasterizer {
+        let loaded_font = self.get_loaded_font(glyph.font_key)?;
+
+        match self.rasterize_glyph(&loaded_font.face, glyph.size, glyph.c) {
+            Ok(glyph) => Ok(glyph),
+            Err(err @ Error::MissingGlyph(_)) => {
+                let fallback_font = self.get_fallback_font(&loaded_font, glyph.c).ok_or(err)?;
+                self.rasterize_glyph(&fallback_font.create_font_face(), glyph.size, glyph.c)
+            }
+            Err(err) => Err(err),
```
```suggestion
            Err(err @ Error::MissingGlyph(_)) => {
                let fallback_font = self.get_fallback_font(&loaded_font, glyph.c).ok_or(err)?;
                self.rasterize_glyph(&fallback_font.create_font_face(), glyph.size, glyph.c)
            },
            result => result,
```

I'm not sure if there isn't an even better way to do this, but I feel like this makes things at least a little nicer.

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -186,29 +276,85 @@ pub enum Error {
+    let mut buf = vec![0u16; LOCALE_NAME_MAX_LENGTH];
+    let len = unsafe { GetUserDefaultLocaleName(buf.as_mut_ptr(), buf.len() as i32) as usize };
+
+    // len includes null byte, we don't need that in Rust strings
```
```suggestion
    // `len` includes null byte, which we don't need in Rust
```

chrisduerr commented on font/src/directwrite/mod.rs:
```diff
@@ -186,29 +276,85 @@ pub enum Error {
+    OsString::from_wide(&buf[..len - 1]).into_string().expect("Locale not valid unicode")
+}
+
+/// Information needed by dwrote's TextAnalysisSource to find a fallback font
+/// for a run of text. See `get_fallback_font`.
```
Usually the first line of documentation on an item should be a short description of what it is. I feel like if that doesn't even fit on a single line, it's probably too verbose.

```suggestion
/// Font fallback information for dwrote's TextAnalysisSource.
```

Do you think this correctly represents the purpose of this struct, or is there some information missing?

davidhewitt commented:
Thanks, have applied suggestions and rebased. Couple of bits of input needed from you to decide how to finish things off.

davidhewitt commented:
That's done 👍 

As there's a lot of commits on this branch, probably worth squashing as part of the merge.

madetara commented:
I just want to thank you all for making it!
Can't wait for it to be released

Thank you!

s-taylor commented:
I'd also love for this to be released, i spent hours puzzling why my powerline icons were not rendering thinking it was something wrong with my fonts. I'll see if i can try build it manually from source but that's likely painful on windows.

jsinkers commented:
I can confirm building from source is not too difficult on Windows and
worth the effort. It seems like fixing font fallback would be a minor
change but I found this to be a significant improvement.

On Wed, 6 May 2020 at 16:39, Simon Taylor <notifications@github.com> wrote:

> I'd also love for this to be released, i spent hours puzzling why my
> powerline icons were not rendering thinking it was something wrong with my
> fonts. I'll see if i can try build it manually from source but that's
> likely painful on windows.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/alacritty/alacritty/pull/3127#issuecomment-624469521>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABX2PQAK6D3DISSLTVVKKMLRQEAZJANCNFSM4J737YHQ>
> .
>

s-taylor commented:
For anyone struggling with this, I was able to resolve it with these steps

* Build Alacritty from source (the steps to this were not clear, but I figured it out, [see notes here](https://gist.github.com/s-taylor/5aa53297cda0a7dd7c3539bcbf952a06))
* Install fira code via choco `choco install firacode`
* Set fira code as my font in the alacritty.yml and set `use_thin_strokes: true`
* Zoom in slightly

I feel like the other font installation methods I was using were not working correctly. Suspect there is more to installing fonts in windows than just double clicking on the font files and choosing install. But I did also try other installation methods such as [this](https://medium.com/@slmeng/how-to-install-powerline-fonts-in-windows-b2eedecace58) which did not work either. Installing fira code through choco seems to have done the magic.
